### ISSUE TITLE
Type is lost when slicing a subclass of recarray

### ISSUE THREAD

**Issue by westphae** [numpy/numpy#6949]
A bug was introduced in Numpy 1.10.2 whereby slicing a subclass of rec.array loses the subclass type and returns a rec.array instead.

Demo code:
```python
In [1]: import numpy as np

In [2]: class C(np.recarray): pass

In [3]: c = C((2,), dtype=[('x', float), ('y', int)])

In [4]: type(c)
Out[4]: __main__.C

In [5]: c2 = c[:]

In [6]: type(c2)
Out[6]: numpy.recarray
```

in 1.10.1 and earlier, the final (correct) output was:
```python
In [6]: type(c2)
Out[6]: __main__.C
```

The offending code is in numpy/core/records.py, line 502 in v1.10.1 which was changed to lines 505-508 in v1.10.1.  Here's the diff:

```python
@@ -499,7 +502,10 @@ class recarray(ndarray):
         # we might also be returning a single element
         if isinstance(obj, ndarray):
             if obj.dtype.fields:
-                return obj.view(dtype=(self.dtype.type, obj.dtype.fields))
+                obj = obj.view(recarray)
+                if issubclass(obj.dtype.type, nt.void):
+                    return obj.view(dtype=(self.dtype.type, obj.dtype))
+                return obj
             else:
                 return obj.view(type=ndarray)
```

System is Arch Linux, python 3.5.1

**Comment by charris** [comment]
Urk. @ahaldane Any chance of getting a quick fix for this that I could add to 1.10.3?

**Comment by charris** [comment]
@westphae I think this may be intended behavior, so will wait to hear from @ahaldane. What sort of problem is it causing you?

**Comment by charris** [comment]
See #6641.

**Comment by westphae** [comment]
Hmm, thanks for the quick response and the links, @charris.  I hadn't been able to find any discussion of this behavior.

My issue is that I have a subclass of `rec.array` that has additional methods defined.  If I take a slice of that subclass, then I no longer am able to call those methods since the class has changed.  I think my issue is different from that in #6641, where the slicing is across fields rather than records.  Arguably, if I sliced my `c` class across some fields, it might no longer be considered type `c` and returning a `rec.array` would make sense.

I do agree with #6641 that older versions of numpy where `type(c[['x']])` returns `numpy.ndarray` is not the expected behavior.

Back to my example: I might expect `c[0]` to return `numpy.record`, and also might expect `c[['x']]` to return `numpy.recarray`, but would expect `c[:]` (or even `c[:0]`) to return an object of type `C`.  (Actually it's the `c[:0]` case that tripped me up initially in 1.10.2--I needed it to preserve my subclass type `C`.)

**Comment by charris** [comment]
So you would like to preserve record subclasses?

**Comment by westphae** [comment]
Yes, under slicing, I would like to preserve rec.array and probably record subclasses.  This would also parallel how subclasses of ndarray are preserved:

```python
In [7]: class D(np.ndarray): pass

In [8]: d = D((2,))

In [9]: type(d)
Out[9]: __main__.D

In [10]: type(d[:])
Out[10]: __main__.D
```

**Comment by ahaldane** [comment]
Hmm, just off the top of my head, it sounds like you're right. Remember that record arrays involve two types of subclassing: The array type is `np.recarray` and the dtype is `np.record`. It does sound like we should try to keep `np.recarray` subclasses. Probably somewhere in the code we need to replace `np.view(type=np.recarray, dtype=(np.record, self.dtype))` by `np.view(type=type(self), dtype=(np.record, self.dtype))`.

I don't have time to look at the code immediately (I've been pretty busy these last few weeks) but I'll try and take a look some evening soon, if no one does so first. When do you want to get 1.10.3 out?

**Comment by charris** [comment]
| When do you want to get 1.10.3 out?

I was shooting for today ;) There is a nasty segfault that needs fixing. I also want to branch 1.11.0 in a week or two and go for a faster release. Sounds to me like 1.11 might be the proper release for this fix. @westphae I know that that would be an inconvenience for you, but we have limited developer time.

**Comment by westphae** [comment]
Understood!  For now I reverted back to 1.10.1; I'll keep an eye out for the fix.  Thanks for the attention!

**Comment by charris** [comment]
Oh, 1.10.1 has a ton of problems also, the most severe (IMHO) being #6532... We can always do a 1.10.4. I expected 1.10.x to be a difficult release because there were a lot of accumulated changes. Needing to make a lot of fixup releases is a bit embarrassing, but that is how it seems to be shaking out.

**Comment by westphae** [comment]
Thanks for the heads-up! Maybe I'll revert to 1.8.2, which is tested for us in production...

I may attempt a patch for this myself, if I can get the time.  Maybe it helps @ahaldane.

**Comment by mhvk** [comment]
I think this is fairly easy -- PR upcoming.

**Comment by mhvk** [comment]
See #6952

**Comment by westphae** [comment]
Awesome!  Thanks.

**Comment by charris** [comment]
Fixed by #6952.
